rust/lance-io/src/object_store/gcs_wrapper.rs

```diff
@@ -0,0 +1,458 @@
+// SPDX-License-Identifier: Apache-2.0
+// SPDX-FileCopyrightText: Copyright The Lance Authors
+
+//! Wrappers around object_store that apply tracing
+
+use std::io;
+use std::ops::Range;
+use std::pin::Pin;
+use std::sync::Arc;
+use std::task::Poll;
+
+use async_trait::async_trait;
+use bytes::Bytes;
+use futures::future::BoxFuture;
+use futures::stream::{BoxStream, FuturesUnordered};
+use futures::{FutureExt, StreamExt};
+use object_store::gcp::GoogleCloudStorage;
+use object_store::multipart::{MultiPartStore, PartId};
+use object_store::path::Path;
+use object_store::{
+    Error as OSError, GetOptions, GetResult, ListResult, MultipartId, ObjectMeta, ObjectStore,
+    PutOptions, PutResult, Result as OSResult,
+};
+use rand::Rng;
+use tokio::io::AsyncWrite;
+
+/// Wrapper around GoogleCloudStorage with a larger maximum upload size.
+///
+/// This will be obsolete once object_store 0.10.0 is released.
+#[derive(Debug)]
+pub struct PatchedGoogleCloudStorage(pub Arc<GoogleCloudStorage>);
+
+impl std::fmt::Display for PatchedGoogleCloudStorage {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "PatchedGoogleCloudStorage({})", self.0)
+    }
+}
+
+#[async_trait]
+impl ObjectStore for PatchedGoogleCloudStorage {
+    async fn put(&self, location: &Path, bytes: Bytes) -> OSResult<PutResult> {
+        self.0.put(location, bytes).await
+    }
+
+    async fn put_opts(
+        &self,
+        location: &Path,
+        bytes: Bytes,
+        opts: PutOptions,
+    ) -> OSResult<PutResult> {
+        self.0.put_opts(location, bytes, opts).await
+    }
+
+    async fn put_multipart(
+        &self,
+        location: &Path,
+    ) -> OSResult<(MultipartId, Box<dyn AsyncWrite + Unpin + Send>)> {
+        // We don't return a real multipart id here. This will be addressed
+        // in object_store 0.10.0.
+        Upload::new(self.0.clone(), location.clone())
+            .map(|upload| (MultipartId::default(), Box::new(upload) as _))
+    }
+
+    async fn abort_multipart(&self, location: &Path, multipart_id: &MultipartId) -> OSResult<()> {
+        MultiPartStore::abort_multipart(self.0.as_ref(), location, multipart_id).await
+    }
```

Comment:
Maybe better to just panic here (or return err) to make it more clear in the future that the reason this fails is because we don't actually return a `MultipartId`?  Just in case this lingers and we start using multipartid at some point in the future to do aborts?